### ISSUE TITLE
pscheduler-api-server.wsgi premature end of script headers

### ISSUE THREAD

**Issue by ajragusa** [perfsonar/pscheduler#631]
We encountered this error from apache ssl_error_log several times, and it prevents things from being scheduled. 
`[error] [client X.X.X.X] Premature end of script headers: pscheduler-api-server.wsgi, referer: https://pas-tst.char.net.internet2.edu/pscheduler/tasks`
I don't see any other relevant logs on the system related to this error

**Comment by mfeit-internet2** [comment]
These are caused by the server cutting a long(er)-running query off at the knees while Apache restarts a WSGI process.  (WSGI is how the Flask framework that runs the API plugs into Apache.)  Apache first asks the process for an orderly shutdown and, after some time, will forcibly kill it if it hasn't exited.

Right now, pScheduler is configured so server processes are restarted after they've handled 5,000 requests.   That seemed like a lot at the time I set it, but some back-of-the-napkin math says that may be too frequent.  Making the number larger is a viable option, and it will be set to 2,500,000 in the upcoming 4.1 release.  That doesn't guarantee that a process being killed won't coincide with a long-running request; it will just make it 500 times rarer.

If you want to experiment with tuning this number on your own systems in the mean time, it's in the first non-comment line of `/etc/httpd/conf.d/pscheduler-api-server.conf`.  Restarting Apache afterward is required.

The restart feature can be turned off entirely, but I'm reluctant to do that since anything that develops a leak would fester and eventually OOM the system.  I think the occasional failed request is a better fate than a slow descent into server failure.

One other thing that can be done is to make the functions in pScheduler that handle HTTP requests retry when the returned status doesn't look like success.  There's a big downside to this, which is that if a server develops a load-related problem, the clients' reaction will only serve to amplify it by a factor of _n_, where _n_ is the number of retries.

**Comment by mfeit-internet2** [comment]
Addendum:  There are some features in mod_wsgi that can be used to make the shutdowns more graceful, but they're not present in the version supplied with EL7.

**Comment by mfeit-internet2** [comment]
**Release Notes:**

Added additional server processes and reduced the frequency of restart to improve performance and reliability.

**Comment by ajragusa** [comment]
Hey Mark

This feels more like a work-around than an actual fix.  Can we re-open this and determine the root cause, even with it killing the application shouldn't this process be really fast and non-seen by end users.  Seems like there is something else going on.

Thanks
A.J.

**Comment by mfeit-internet2** [comment]
> ...even with it killing the application shouldn't this process be really fast and non-seen by end users. Seems like there is something else going on.

There are some endpoints in pScheduler's REST API that are long-running because they don't return a result until one is available.  It's preferable to do this in the server instead of having external clients hammering away at the server with repeated HTTP requests.  These endpoints have a hard limit of 30 seconds.  Clients can ask for less but not more.

With that background, this is exactly what's going on, as verified by in the mod_wsgi source code:

 - Because the WSGI process is multi-threaded, it may successfully finish a short-lived request that puts it across the restart threshold while a longer-running request is still running.
 - The Apache side of mod_wsgi sends the WSGI process the graceful shutdown signal.
 - The WSGI process starts a thread that sleeps through the graceful shutdown timeout and then calls `exit()`.
 - Meanwhile, the WSGI process attempts to clean up after itself by calling `join()` against all of its outstanding threads.  If that process doesn't finish, the threads are killed off when the thread above terminates the process.

Here's the problem:  RHEL 7 (and, by extension CentOS 7) supplies mod_wsgi 3.x, which is what was stable what it was released.  That version has a graceful shutdown timeout that's hard-wired to five seconds, so I can't guarantee that a query will never be interrupted.  I wasn't aware of this limitation when I first wrote the code, so that one's on me.

There are knobs in mod_wsgi 4.x that will let me configure this so it works properly.  The upgrade should be easy, and I think we should do it  We are, however, very close enough to putting 4.1 into beta, and I'm not sanguine about a change to pScheduler's underpinnings without some time in the test bed to make sure it doesn't break anything else.

If you think we should hold the release for this (and I have no objection to that), take it up with @arlake228 and the other leads.  Do keep in mind that the changes I made (a) were needed anyway because we only had one WSGI process with 15 threads in it and (b) will reduce the incidence of this by a factor of at least 5,000.

PAS and WLCG are the two meshes I know of that are pushing pScheduler the hardest.  The development team doesn't have the resources to do do an equivalent test bed, so unfortunately, we have to find out about these things out in the field.
